src/main/java/io/cdap/plugin/batch/sink/RecordToBSONWritableTransformer.java

```diff
@@ -65,124 +68,105 @@ public RecordToBSONWritableTransformer(@Nullable String idFieldName) {
-    Schema recordWithIdSchema = Schema.recordOf(recordName, recordWithIdFields);
-    StructuredRecord.Builder builder = StructuredRecord.builder(recordWithIdSchema);
-    for (Schema.Field field : original.getSchema().getFields()) {
-      if (idFieldName.equals(field.getName())) {
-        builder.set(DEFAULT_ID_FIELD_NAME, original.get(idFieldName));
-        continue;
-      }
-
-      builder.set(field.getName(), original.get(field.getName()));
-    }
-
-    return builder.build();
+    extracted.removeField(idFieldName);
+    extracted.put(DEFAULT_ID_FIELD_NAME, extractObjectId(idFieldName, record.get(idFieldName)));
+    return new BSONWritable(extracted);
   }
 
-
-  private DBObject extractRecord(Object value, @Nullable String idFieldName) {
-    if (value == null) {
+  private DBObject extractRecord(@Nullable String fieldName, StructuredRecord record) {
+    if (record == null) {
       // Return 'null' value as it is
       return null;
     }
-    StructuredRecord record = (StructuredRecord) value;
     List<Schema.Field> fields = Objects.requireNonNull(record.getSchema().getFields(), "Schema fields cannot be empty");
     BasicDBObjectBuilder bsonBuilder = BasicDBObjectBuilder.start();
     for (Schema.Field field : fields) {
-      Schema.LogicalType fieldLogicalType = field.getSchema().isNullable()
-        ? field.getSchema().getNonNullable().getLogicalType()
-        : field.getSchema().getLogicalType();
+      // Use full field name for nested records to construct meaningful errors messages.
+      // Full field names will be in the following format: 'record_field_name.nested_record_field_name'
+      String fullFieldName = fieldName != null ? String.format("%s.%s", fieldName, field.getName()) : field.getName();
+      Schema nonNullableSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable()
+        : field.getSchema();
+      bsonBuilder.add(field.getName(), extractValue(fullFieldName, record.get(field.getName()), nonNullableSchema));
+    }
+    return bsonBuilder.get();
+  }
 
-      if (!Strings.isNullOrEmpty(idFieldName) && idFieldName.equals(field.getName())) {
-        bsonBuilder.add(idFieldName, extractObjectId(record.get(field.getName())));
-        continue;
+  private ObjectId extractObjectId(String fieldName, Object value) {
+    try {
+      if (value instanceof byte[]) {
+        return new ObjectId((byte[]) value);
+      } else if (value instanceof ByteBuffer) {
+        return new ObjectId((ByteBuffer) value);
+      } else if (value instanceof String) {
+        return new ObjectId((String) value);
       }
 
-      if (fieldLogicalType == null) {
-        bsonBuilder.add(field.getName(), extractValue(record.get(field.getName()), field.getSchema()));
-        continue;
-      }
+      throw new UnexpectedFormatException(
+        String.format("Identifier field '%s' is expected to be of type '%s', '%s', '%s', but found a '%s'.", fieldName,
```

Comment:
the error message should use types from the schema and not the java class name. So something like:

Identified field '%s' is expected to be of type 'bytes' or 'string', but found a '%s'

where the type at the end is from the schema and not a class name.

Reply:
Done.